openslides/assignments/views.py

```diff
@@ -242,6 +243,134 @@ def get_serializer_class(self):
+        """
+        if not request.user.has_perm('assignments.can_nominate_other'):
+            self.permission_denied(request)
+        user = self.get_other_user(request)
+        assignment = self.get_object()
+        if assignment.is_elected(user):
+            raise ValidationError({'detail': _('User %s is already elected.') % user})
+        if request.method == 'POST':
+            message = self.nominate_other(request, user, assignment)
+        else:
+            # request.method == 'DELETE'
+            message = self.delete_other(request, user, assignment)
+        return Response({'detail': message})
+
+    def nominate_other(self, request, user, assignment):
+        if assignment.phase == assignment.PHASE_FINISHED:
+            raise ValidationError({'detail': _('You can not nominate someone to this election because it is finished.')})
+        if assignment.phase == assignment.PHASE_VOTING and not request.user.has_perm('assignments.can_manage'):
+            # To nominate other during voting you have to be a manager.
+            self.permission_denied(request)
+        if assignment.is_candidate(user):
+            raise ValidationError({'detail': _('User %s is already a candidate of this election.') % user})
+        if assignment.is_blocked(user) and not request.user.has_perm('assignments.can_manage'):
+            raise ValidationError({'detail': _('User %s does not want to be an candidate.') % user})
+        assignment.set_candidate(user)
+        return _('User %s was nominated successfully.') % user
+
+    def delete_other(self, request, user, assignment):
+        # To delete candidature status you have to be a manager.
+        if not request.user.has_perm('assignments.can_manage'):
+            self.permission_denied(request)
+        if assignment.phase == assignment.PHASE_FINISHED:
+            raise ValidationError({'detail': _('You can not delete someones candidature to this election because it is finished.')})
+        if not assignment.is_candidate(user) and not assignment.is_blocked(user):
+            raise ValidationError({'detail': _('User %s has no status in this election.') % user})
+        assignment.delete_related_user(user)
+        return _('Candidate %s was withdrawn/unblocked successfully.') % user
+
+    @detail_route(methods=['post', 'delete'])
+    def mark_elected(self, request, pk=None):
+        """
+        View to mark other users as elected (POST) undo this (DELETE). The
+        client has to send {'user': <id>}.
+        """
+        if not request.user.has_perm('assignments.can_manage'):
+            self.permission_denied(request)
+        user = self.get_other_user(request)
+        assignment = self.get_object()
+        if request.method == 'POST':
+            if not assignment.is_candidate(user):
+                raise ValidationError({'detail': _('User %s is not a candidate of this election.') % user})
+            assignment.set_elected(user)
```

Comment:
The python way is to call a method and if it fails to get the Exception with try-except. So I would prefer that "assignment.set_elected()" would raise an error, if the user is not a candidate.

But the method "assignment.set_elected" uses the django method "update_or_create". So either, we say, that if 'set_elected' is called on a new user, the new user has to become a related user of the assignment which is elected, or "assignment.set_elected" should use the django method "update" instead.

I would prefer that if the user is not a candidate, he/she can nevertheless marked as elected.